### ISSUE TITLE
[angular2] Compile error with Angular 4 AoT providing Configuration

### ISSUE THREAD

**Issue by kernwig** [swagger-api/swagger-codegen#5349]
##### Description

Error when trying to import `ApiModule.forConfig(myConfiguration)` from a const Configuration object.

Error: Error encountered resolving symbol values statically. Calling function 'Configuration', function calls are not supported. Consider replacing the function or lambda with a reference to an exported function

##### Swagger-codegen version

2.3.0

##### Related issues

Information about this problem: https://github.com/angular/angular-cli/issues/3707

##### Suggest a Fix

```
--- a/modules/swagger-codegen/src/main/resources/typescript-angular2/api.module.mustache
+++ b/modules/swagger-codegen/src/main/resources/typescript-angular2/api.module.mustache
@@ -16,10 +16,10 @@ import { {{classname}} } from './{{importPath}}';
   providers:    [ {{#apiInfo}}{{#apis}}{{classname}}{{#hasMore}}, {{/hasMore}}{{/apis}}{{/apiInfo}} ]
 })
 export class ApiModule {
-    public static forConfig(configuration: Configuration): ModuleWithProviders {
+    public static forConfig(configurationFactory: () => Configuration): ModuleWithProviders {
         return {
             ngModule: ApiModule,
-            providers: [ {provide: Configuration, useValue: configuration}]
+            providers: [ {provide: Configuration, useFactory: configurationFactory}]
         }
     }
 }
```

**Comment by wing328** [comment]
@kernwig thanks for reporting the issue. Please file a PR so that we can review the fix.

**Comment by wing328** [comment]
If this change is for Angular 4 only, please add an option "angular4" for the `library` option.

**Comment by kernwig** [comment]
I don't actually know. I found AoT unusable in Angular 2. Just trying it again after upgrading to Angular 4, but the linked report is from last December, so it is probably true with Angular 2 also. The fix will work in both for sure. I don't know what `library` is. Heading out on vacation momentarily, so it'll be a while before I can pursue this further.

**Comment by wing328** [comment]
For other API clients (e.g. Java), we provide support for different HTTP libraries via the `library` option:
```
	library
	    library template (sub-template) to use (Default: okhttp-gson)
	        jersey1 - HTTP client: Jersey client 1.19.1. JSON processing: Jackson 2.7.0. Enable Java6 support using '-DsupportJava6=true'. Enable gzip request encoding using '-DuseGzipFeature=true'.
	        feign - HTTP client: OpenFeign 9.4.0. JSON processing: Jackson 2.8.7
	        jersey2 - HTTP client: Jersey client 2.22.2. JSON processing: Jackson 2.7.0
	        okhttp-gson - HTTP client: OkHttp 2.7.5. JSON processing: Gson 2.6.2. Enable Parcelable modles on Android using '-DparcelableModel=true'. Enable gzip request encoding using '-DuseGzipFeature=true'.
	        retrofit - HTTP client: OkHttp 2.7.5. JSON processing: Gson 2.3.1 (Retrofit 1.9.0). IMPORTANT NOTE: retrofit1.x is no longer actively maintained so please upgrade to 'retrofit2' instead.
	        retrofit2 - HTTP client: OkHttp 3.2.0. JSON processing: Gson 2.6.1 (Retrofit 2.0.2). Enable the RxJava adapter using '-DuseRxJava[2]=true'. (RxJava 1.x or 2.x)
```
For Angular2, we can do something similar to support newer version for Angular (instead of creating another generator)

Enjoy the vacation!

**Comment by kernwig** [comment]
Ah. Not needed as of yet. I haven't seen anything in swagger-codegen that matters between Angular 2 vs 4. I just happened to try this in Angular 4. I included it in the title because I cannot confirm that's what I can test with at this point.

I can say that my suggested fix works, and should be fine for Angular 2 as well. Whether or not it's the best fix, I don't know. I see that value provider is used in Angular's RouterModule and that works with AoT, so it seems like swagger's current output should work too. Thus there is probably another option that I'm not seeing.

**Comment by kenisteward** [comment]
Is there a PR for this?

**Comment by wing328** [comment]
@kenisteward not that I'm aware of. If you want to work on the enhancement, please reply to let us know.

**Comment by kernwig** [comment]
@wing328 Can you look at my pull request? I'm going to be handing my project over to a client and I'd really like for the 2.3.0 release to be usable by them, rather than continuing to custom build.

**Comment by wing328** [comment]
Closed via #5560
